runtime/src/main/kotlin/edu/cornell/cs/apl/viaduct/backend/PlaintextProtocolInterpreter.kt

```diff
@@ -218,37 +218,64 @@ class PlaintextProtocolInterpreter(
                         }
                     }
 
-                    assert(cleartextValue != null)
+                    if (cleartextValue == null) {
+                        throw ViaductInterpreterError("Plaintext: received null value")
+                    }
 
-                    // check for equivocation
-                    /*
-                    val recvHosts: Set<Host> =
+                    // calculate set of hosts with whom [this.host] needs to check for equivocation
+                    val hostsToCheckWith: Set<Host> =
                         events
-                            .filter { event -> event.recv.id == Plaintext.INPUT && event.send.host != event.recv.host }
+                            .filter { event ->
+                                // remove events where receiving host is not receiving plaintext data
+                                event.recv.id == Plaintext.INPUT &&
+
+                                    // remove events where a host is sending data to themselves
+                                    event.send.host != event.recv.host &&
+
+                                    // remove events where [this.host] is the sender of the data
+                                    event.send.host != this.host
+                            }
+
+                            // of events matching above criteria, get set of data receivers
                             .map { event -> event.recv.host }
+
+                            // remove [this.host] from the set of hosts with whom [this.host] needs to
+                            // check for equivocation
                             .filter { host -> host != this.host }
                             .toSet()
 
-                    for (recvHost in recvHosts) {
+                    if (hostsToCheckWith.isNotEmpty()) {
+                        logger.trace {
+                            "host: " + this.host.asDocument.print() + " checks for equivocation with: " +
+                                hostsToCheckWith.fold("") { acc, host -> host.asDocument.print() + ' ' + acc }
```

Comment:
Instead of `fold`, use `List<PrettyPrintable>.joined`. You should also sort the hosts using `.sorted()`.